### ISSUE TITLE
BoTorch: Programmable Bayesian Optimization in PyTorch

### ISSUE THREAD

**Issue by sile** [sile/readings#58]
URL
-----

https://arxiv.org/abs/1910.06403

Abstract
----------

> Bayesian optimization provides sample-efficient global optimization for a broad range of applications, including automatic machine learning, molecular chemistry, and experimental design. We introduce BoTorch, a modern programming framework for Bayesian optimization. Enabled by Monte-Carlo (MC) acquisition functions and auto-differentiation, BoTorch's modular design facilitates flexible specification and optimization of probabilistic models written in PyTorch, radically simplifying implementation of novel acquisition functions. Our MC approach is made practical by a distinctive algorithmic foundation that leverages fast predictive distributions and hardware acceleration. In experiments, we demonstrate the improved sample efficiency of BoTorch relative to other popular libraries. BoTorch is open source and available at this https URL. 

Table of Contents
--------------------

- (1) INTRODUCTION
  - Systems
  - Methodology
- (2) BACKGROUND AND RELATED WORK
- (3) BOTORCH ARCHITECTURE
  - Model
  - Objective
  - Acquisition function
  - Optimizer
- (4) ACQUISTION FUNCTIONS
  - (4-1) Analytic Acquisition Functions
  - (4-2) Monte-Carlo Acquisition Functions
    - Arbitrary posteriors
    - Arbitrary objectives and constraints
    - Parallel evaluation
    - Asynchronous evaluation
  - (4-3) Computing Gradients
  - (4-4) Variance Reduction via qMC Sampling
- (5) API ABSTRACTIONS
  - Model
  - Posterior
  - MCSampler
  - MCObjective
- (6) IMPLEMENTATION EXAMPLES
  - (6-1) Composite Objectives
  - (6-2) Parallel Noisy Expected Improvement
  - (6-3) Active Learning
- (7) OPTIMIZING ACQUISITION FUNCTIONS
- (8) EXPLOITING PARALLELISM AND HARDWARE ACCELERATION
  - (8-1) Batch Evaluation
  - (8-2) Fast Posterior Evaluation
- (9) ONE-SHOT OPTIMIZATION of KG
- (10) OPTIMIZATION EXPERIMENTS
  - (10-1) Synthetic Test Functions
  - (10-2) Computational Scaling
  - (10-3) Hyperparameter Optimization
    - (10-3-1) DQN and Cartpole
    - (10-3-2) Neural Network Surrogate
    - (10-3-3) Stochastic Weight Averaging for CIFAR-10
- (11) DISCUSSION AND OUTLOOKb
- REFERENCES

Appendix:
- (A) REVIEW OF OTHER BO PACKAGES
  - Spearmint
  - GPyOpt
  - Cornell-MOE
  - RoBO
  - Emukit
  - Dragonfly
- (B) OPTIMIZATION WITH FIXED SAMPLES
- (C) ADDITIONAL EMPIRICAL RESULTS
  - (C-1) Synthetic Functions
  - (C-2) Constrained Bayesian Optimization
  - (C-3) Stochastic Weight Averaging on CIFAR-10
- (D) BATCH INITIALIZATION FOR MULTI-START OPTIMIZATION
- (E) EXAPMLES
  - (E-1) Composite Objectives
  - (E-2) Active Learning
  - (E-3) Generalized UCB
  - (E-4) Full Code Examples

BibTeX
------

```bibtex
@article{balandat2019botorch,
  title={BoTorch: Programmable Bayesian Optimization in PyTorch},
  author={Balandat, Maximilian and Karrer, Brian and Jiang, Daniel R and Daulton, Samuel and Letham, Benjamin and Wilson, Andrew Gordon and Bakshy, Eytan},
  journal={arXiv preprint arXiv:1910.06403},
  year={2019}
}
```

See Also
----------

- GitHub: https://github.com/pytorch/botorch
